wallet-new/src/Cardano/Wallet/API/V1/Generic.hs

```diff
@@ -1,166 +1,120 @@
-                in hsequence $ hcliftA pf mkVal vals
-parseJsendValues (JsonRecord tag fields) =
-    unJsendValue tag $
-    withObject "Object" $ \o ->
-        let getField :: FromJSON a => K String a -> Parser a
-            getField (K name) = o .: toText name
-        in hsequence $ hcliftA pf getField fields
-
--- | Helper function which makes a JSON value in JSend format
--- from a constructor tag and object with constructor's arguments
-jsendValue :: ResponseStatus -> Tag -> Value -> K Value a
-jsendValue _ NoTag v   = K v
-jsendValue rs (Tag t) v = K $ Object $
-    HM.fromList [ ("message", String $ toText t)
-                , ("diagnostic", v)
-                , ("status", toJSON rs)
-                ]
-
--- | Helper function to parse value in JSend format if desired constructor
--- is known.
-unJsendValue :: Tag -> (Value -> Parser a) -> Value -> Parser a
-unJsendValue NoTag f = f
-unJsendValue (Tag n) f = withObject ("Expected JSend object with message `" <> n <> "`") $
-    \o -> do
-        msg <- o .: "message"
-        guard $ n == msg
-        val <- o .: "diagnostic"
-        f val
 
 --
--- Misc
+-- INTERNALS
 --
 
 gconsInfos
-    :: forall a. (HasDatatypeInfo a)
-    => Proxy a -> NP ConstructorInfo (Code a)
-gconsInfos pa = case datatypeInfo pa of
-    Newtype _ _ conInfo -> conInfo :* Nil
-    ADT _ _ consInfo    -> consInfo
+    :: forall a. (SOP.HasDatatypeInfo a)
+    => Proxy a
+    -> SOP.NP SOP.ConstructorInfo (SOP.Code a)
+gconsInfos pa = case SOP.datatypeInfo pa of
+    SOP.Newtype _ _ conInfo -> conInfo SOP.:* SOP.Nil
+    SOP.ADT _ _ consInfo    -> consInfo
+
+
+-- | This class helps us define generically errors JSON instances without
+-- relying on partial field in records.
+-- This is used to encode the diagnostic object of an error, as a singleton
+-- with field 'Text' whenever there's one.
+--
+-- NOTE: We haven't defined instances for everything because we do not want to
+-- suppport all kind of error structures, but only sums with a unary or
+-- nullary constructors.
+class GDiagnosticToJSON (f :: * -> *) where
+    gDiagnosticToJSON :: Text -> f a -> Value
 
-gconsNames
-    :: forall a. (HasDatatypeInfo a, SListI (Code a))
-    => Proxy a -> [Text]
-gconsNames =
-  map toText . hcollapse . hliftA (K . constructorName) . gconsInfos
+instance (GDiagnosticToJSON f) => GDiagnosticToJSON (M1 i c f) where
+    gDiagnosticToJSON k (M1 f) = gDiagnosticToJSON k f
 
-gconsName
-  :: forall a. (Generic a, HasDatatypeInfo a)
-  => a -> Text
-gconsName a =
-  gconsNames (Proxy @a) !! hindex (from a)
+instance (GDiagnosticToJSON f, GDiagnosticToJSON g) => GDiagnosticToJSON (f :+: g) where
+    gDiagnosticToJSON k (L1 f) = gDiagnosticToJSON k f
+    gDiagnosticToJSON k (R1 g) = gDiagnosticToJSON k g
+
+instance (ToJSON c) => GDiagnosticToJSON (K1 i c) where
+    gDiagnosticToJSON k (K1 c) = object [ k .= c ]
+
+instance GDiagnosticToJSON U1 where
+    gDiagnosticToJSON _ _ = object mempty
```

Comment:
Gotta trust the tests on that one I guess. 
The diff is better visualized when looking at commits: 

- first, the old implementation using json-sop was removed (1bd80bebd0a1d69bd70a5fec851783df45874a99)
- and later, a new implementation which doesn't rely on partial field accessor was done (72a608e218995f70f473145b4f32f3704756ca04)